saleor/discount/utils.py

```diff
@@ -610,20 +612,44 @@ def _get_discount_amount(
         or discounted_price_amount is None
         or price_amount == discounted_price_amount
     ):
-        return Decimal("0.0")
+        return False
 
-    unit_discount = price_amount - discounted_price_amount
-    return unit_discount * line_quantity
+    return True
 
 
 def _get_rule_discount_amount(
-    variant_listing_promotion_rule: Optional["VariantChannelListingPromotionRule"],
-    line_quantity: int,
+    line: Union["CheckoutLine", "OrderLine"],
+    rule_info: "VariantPromotionRuleInfo",
+    channel: "Channel",
 ) -> Decimal:
+    """Calculate the discount amount for catalogue promotion rule.
+
+    When the line has overridden price, the discount is applied on the
+    new overridden base price.
+    """
+    variant_listing_promotion_rule = rule_info.variant_listing_promotion_rule
     if not variant_listing_promotion_rule:
         return Decimal("0.0")
-    discount_amount = variant_listing_promotion_rule.discount_amount
-    return discount_amount * line_quantity
+
+    if isinstance(line, CheckoutLine):
+        price_override = (
+            Money(line.price_override, channel.currency_code)
+            if line.price_override is not None
+            else None
+        )
+    else:
+        price_override = (
+            line.undiscounted_base_unit_price if line.is_price_overridden else None
+        )
+
+    if price_override:
```

Comment:
Shouldn't be:
```suggestion
    if price_override is None:
```
The override price can be Decimal(0)

Reply:
Having test for that case would be also good